packages/ember-metal/lib/mixin.js

```diff
@@ -364,6 +364,12 @@ function applyMixin(obj, mixins, partial) {
       value = followed.value;
     }
 
+    if (desc && desc.isDescriptor) {
+      m.writeDescriptor(key, desc);
+    } else if (m.peekDescriptor(key)) {
```

Comment:
If we wanted to avoid traversing the inherited meta's for each key that is merged, we could use `m.forEachDescriptor` once (outside of this loop) and change this guard to check if the key being added is in that listing.

Something like:

```js
let knownDescriptors = [];
m.forEachDescriptor(name => knownDescriptors.push(name));

// ...snip...
  mergeMixins(mixins, m, descs, values, obj, keys);

  for (let i = 0; i < keys.length; i++) {
    // ...snip...
    if (desc && desc.isDescriptor) {
      m.writeDescriptor(key, desc);
    } else if (knownDescriptors.indexOf(key) > -1) {
      m.resetDescriptor(key);
    }
```

Thoughts?